### DISCUSSION TITLE
register_dwi_to_template

### DISCUSSION THREAD

**Discussion by override641** [dipy/dipy#2661] [3. Community Lounge]
Hi folks, 
I am pretty new do dipy, so please forgive if I am asking something obvious. I coregistered B0 to T1, now I am trying to coregister the DWI to B0 using register_dwi_to_template. Once it's done I save it to nifti, however i seem to lose the 4th dimension of the original DWI file. Am I doing something wrong? 
Grateful for any advice
Ivo

**Comment by arokem** [comment]
Hi Ivo - hard to say without seeing your code. Would you mind pasting it here?

**Reply by override641** [reply]
Thank you for your quick reply.
Please see below:

```python
import numpy as np
from dipy.viz import regtools
from dipy.data import fetch_stanford_hardi
from dipy.io.image import load_nifti, save_nifti
from dipy.align.imaffine import (transform_centers_of_mass,
                                 AffineMap,
                                 MutualInformationMetric,
                                 AffineRegistration)
from dipy.align import affine_registration, register_dwi_to_template
from dipy import align
files, folder = fetch_stanford_hardi()
static_data, static_affine, static_img = load_nifti('/Users/override641/Desktop/Reg/registeredB0.nii',
                                            return_img=True)

print(np.shape(static_data))
static=static_data
#static = np.squeeze(static_data)[..., 0]
static_grid2world = static_affine

moving_data, moving_affine, moving_img = load_nifti('/Users/override641/Desktop/Reg/dwi.nii',
                                            return_img=True)

moving = moving_data
moving_grid2world = moving_affine
#moving = np.squeeze(moving_data)[..., 0]
identity = np.eye(4)
print(np.shape(static))
print(np.shape(moving))

nbins = 32
sampling_prop = None
metric = MutualInformationMetric(nbins, sampling_prop)
level_iters = [10, 10, 10]
sigmas = [3.0, 1.0, 0.0]
factors = [4, 2, 1]
pipeline = ["center_of_mass", "translation", "rigid", "affine"]

xformed_dwi, reg_affine = register_dwi_to_template(
    dwi=moving_img,
    gtab=('/Users/override641/Desktop/Reg/dwi.bval',
          '/Users/override641/Desktop/Reg/dwi.bvec'),
    template=static_img,
    reg_method="aff",
    nbins=32,
    metric='MI',
    pipeline=pipeline,
    level_iters=level_iters,
    sigmas=sigmas,
    factors=factors)

print(np.shape(xformed_dwi))
print(np.shape(reg_affine))
save_nifti('/Users/override641/Desktop/Reg/registered_dwi.nii',xformed_dwi, static_affine)
save_nifti('/Users/override641/Desktop/Reg/registered_dwi1.nii',moving_data, reg_affine)
```

**Comment by arokem** [comment]
Hi @override641 : I have been trying to reproduce this error, but I don't have your data, so I've run into some issues even before the error that you are seeing. What is the shape of the various source data? That is, `static_data`, `moving_data`. And what is the number of bvals/bvecs?

**Comment by override641** [comment]
Thank you so much for replying. Does the dwi_to_template rotate the vectors? 
I am trying to do tumor analysis by registering the b0 to t1 and then dwi(multi/shell epi) to b0. Any other idea?
Moving is 100 x 100 x 40 x 70, static 256 x 256 x 180. Bval/bvecs are attached.

Thank you so much for helping me out!
Ivo

**Reply by arokem** [reply]
`register_dwi_to_template` does not rotate the b-vectors. Also, you'd want to be careful using it to transform DWI data, because it applies a non-linear transformation, which means you'd need to rotate the b-vectors by some different amount in every voxel (!). We usually use this function to move data in the template space to the DWI data (i.e., using the backward transform of the mapping object that is returned), and not vice versa.  

For your use-case, I'd reccommend using `affine_registration`. This also does not rotate the b-vectors, but you can use the result to rotate the b-vectors with the `dipy.core.gradients.reorient_bvecs` function.

**Comment by override641** [comment]
Thank you for your reply. One more question - I did register T1 to B0, then B0-DTI and got the reg_affine.   Xformed (output of affine_transform) is only 3D and not 4D as I assumed. Which function do you use to apply the affine transformation to the 4D DWI(DTI) with the reg_affine I got from AffineRegistration?
The vector rotation seemed to have worked.
Thank you so much

**Reply by arokem** [reply]
You can use [`resample`](https://github.com/dipy/dipy/blob/master/dipy/align/_public.py#L291) for that
